### ISSUE TITLE
Unable to reproduce the result in Subword Evenness paper

### ISSUE THREAD

**Issue by kdy20401** [olgapelloni/subword_evenness#1]
Hello Ms Olga,

Firstly, thanks for writing such an interesting paper and sharing codes about the paper.

The reason i submit this issue is that I am having some trouble while reproducing SuE results using the scripts in this repository.
I would really appreciate it if you could provide some helps to me. There are mainly two difficulties.

1. In the count_plot_sue.R file (which is on the measures_scripts/sue/ directory), I cannot execute 
freqs <- add.noise(freqs, 0.0001, "house", 2) this code line because there is no reference to add.noise function.
I think it is not the function R library provides in default. How can I use this add.noise function? And why this process of adding noise the frequency to the elements of dataframe is necessary? 

2. When I calculate the SuE score(an angle) using english_full.txt you provide in this repository, I got an angle of 54.838 degrees(because I cannot execute add.noise function, this is the result of only 1 iteration of for loop). However, the SuE score of English using TeDDi full is 114.54 in this paper.
Instead, I personally make an English full txt corpus using all English TeDDi samples, by concatenating all txt files of all genres and applying some pre-processing. After applying pre-processing to this corpus, the final English txt corpus is like below.

<eng_full.txt>
Illustrated
by
ASHMAN
This
etext
was
produced
...  
(each line contains only one word, and each word consists of only the alphabet)

Then I calculate the SuE score using this eng_fill.txt, I got an angle of 117.686 degrees which is almost the same as the 114.54 which is specified in the paper. In summary, I am curious what is the right way of constructing the full.txt file.

I would be really appreciate if you take time to read this issue and provide some help.

Thank you for reading this long post. Have a nice weekend.

**Comment by olgapelloni** [comment]
Hello @kdy20401,

Thank you for your interest in my code and paper.

1. The function add.noise comes from the library FunChisq https://www.rdocumentation.org/packages/FunChisq/versions/2.5.2/topics/add.noise, in the header of the script count_plot_sue.R there are import statements for all the required libraries. So I guess you just need to install the library FunChisq. I have updated now the script, you can uncomment the lines in the header for installing the required packages.

The measure SuE is sensible to the position of the point with the max(x), that's why introduction of noise helps to get a more reliable value of the angle. During each iteration the max(x) might differ slightly, so the final value of SuE is averaged over 100 iterations with noise. This is not obligatory, but it will provide more reliable results.

2. The *.txt files provided here are intended to use for unsupervised training of LMs. In order to calculate SuE, you need first to run the script segment_count_ui.py over the files ending in _full.txt, which will split the tokens into subwords, count the length of words, count the unevenness index of words (needed for SuE), and create a *.csv output. This csv table serves as input for the script count_plot_sue.R

Hope this helps you to solve the issues. I'll keep in mind that the readme file should be clarified in the future.

**Comment by kdy20401** [comment]
Thanks for answering @olgapelloni!

I could solve the first issue by installing FunChisq package. Thanks for correcting my mistake caused by my unfamiliarity with R.

But, I still have a one question left. I think I totally understand the process of calculating SuE.
(Run the script segment_count_ui.py over the file that ends with _full.txt, which eventually creates the *.csv output. And Use this output file as an input for the count_plot_sue.R script.)

As I mentioned in the first post in this issue session, when I calculate SuE using eng_full.txt you uploaded in the git repository, the value of angle of degree I got (54.82, 54.93, etc.) was far different from the value (118.04 which is at TeDDi 1 mio in Table 7) specified in the paper.

When i see the contents of the english_full.txt for example,

<english_full.txt>
Previously on Underground...
The longer we stand out here fighting', the less chance we got of gettin' Rosalee back.
... (the rest is omitted)

As you can see the above example, each line comprises of a sentence, not a word.
What I conclude after analyzing the script segment_count_ui.py, each line of  *_full.txt file should only contain a single word.
If there are multiple words in one line, that is a sentence, code does not work as you intended I think.

For example, when I executed the script segment_count_ui.py using the english_full.txt file, the result .csv file I got is like below.

word_split	segments_lengths	word_length	index	variance	word	file	genre	language
P|re|vi|ou|s|ly on Und|er|gr|oun|d|...	[1, 2, 2, 2, 1, 9, 2, 2, 3, 1, 3]	28	8	5.072727272727273	Previously on Underground...	eng_full.txt	na	eng
The l|on|g|er we st|and out h|ere f|igh|t|ing|'|, the l|es|s ch|an|ce we go|t of g|et|t|in|' R|os|al|e|e b|ac|k.	[5, 2, 1, 8, 9, 5, 3, 1, 3, 1, 7, 2, 4, 2, 8, 6, 2, 1, 2, 3, 2, 2, 1, 3, 2, 2]	87	8	5.835384615384615	The longer we stand out here fighting', the less chance we got of gettin' Rosalee back.	eng_full.txt	na	eng
....

The value of word_length from the first sentence (Previously on Underground...) is 28.
This is a sentence length not a word length. I think you do not intend to get the sentence length. (Am I right?)

And after processing the english_full.txt as I thought, the SuE score(104.885, 105.36, 112.42, etc) becomes close to the value in the paper. That is why I think it is the right way to construct each line in *_full.txt file as a single word not a sentence.

In short, my question is that
**Is it correct that each line of the .txt file can contain multiple of words? rather than a single word?**
This is only my personal opinion, so I would appreciate it if you could check if this is correct.

Thank you for your reply this weekend. Have a good rest of the weekend.

**Comment by olgapelloni** [comment]
Sorry for misunderstanding, @kdy20401 . 
Indeed, you're right, the files *_full.txt have to undergo preprocessing so that there's *only one* word per line. This is important for segmenting with the BPE algorithm and then yes, of course, in the csv table it has to be just one word per line as well. In the script segment_count_ui.py there were two functions for preprocessing, but for some reason I forgot to call them in the main program (probably forgot while cleaning the experimental scripts and preparing them for publishing here). 

I have committed now some changes to that script, so that it runs preprocessing before segmenting words and counting UI; hope it works fine now. Thanks a lot for spotting this error in my code.
